data/connection.go

```diff
@@ -326,3 +360,60 @@ func (c *PostgresSQL) DeleteOrder(userID int, orderID int) error {
 
 	return nil
 }
+
+// CreateCoffee creates a new coffee
+func (c *PostgresSQL) CreateCoffee(coffee model.Coffee) (model.Coffee, error) {
+	m := model.Coffee{}
+
+	rows, err := c.db.NamedQuery(
+		`INSERT INTO coffees (name, teaser, description, price, image, created_at, updated_at) 
+		VALUES(:name, :teaser, :description, :price, :image, now(), now()) 
+		RETURNING id;`, map[string]interface{}{
+			"name":        coffee.Name,
+			"teaser":      coffee.Teaser,
+			"description": coffee.Description,
+			"price":       coffee.Price,
+			"image":       coffee.Image,
+		})
+	if err != nil {
+		return m, err
+	}
+	defer rows.Close()
+
+	if rows.Next() {
+		err := rows.StructScan(&m)
+		if err != nil {
+			return m, err
+		}
+	}
+
+	return m, nil
+}
+
+// CreateCoffeeIngredient creates a new coffee ingredient
+func (c *PostgresSQL) CreateCoffeeIngredient(coffee model.Coffee, ingredient model.Ingredient) (model.CoffeeIngredients, error) {
+	i := model.CoffeeIngredients{}
+
+	rows, err := c.db.NamedQuery(
```

Comment:
In this current implementation, you can only create new coffee ingredients and don't check for duplications. Also, this assumes will keep previous ingredients.

This present three edge cases:
1. If you hit this endpoint with coffee and milk. Then the second time you hit this endpoint with coffee, milk and sugar. It will error (or should error) since PackerCoffee already has coffee and milk.
2. If you hit this endpoint with coffee and milk. Then the second time you hit this endpoint with sugar. This will return coffee, milk and sugar.
3. If you hit this endpoint with coffee and milk. You're unable to update the quantities of milk and sugar.

---

One way to address this is invalidate all `coffee_ingredients` for that `coffee_id` first before you insert new ones. This solve all three edge cases, but you have to send the full ingredient list every single time.

You can also [upsert](https://www.postgresqltutorial.com/postgresql-upsert/) the ingredients. To remove ingredients, you would just send in quantity 0. You would need to update the ingredients to ignore ingredients with 0 quantity tho.

Reply:
+1 on the upsert approach.  Also, here's a link to the SQL schema. Hopefully, it can help. If you want to have a zoom or something let me know.  

https://github.com/hashicorp-demoapp/product-api-go/blob/4a3934f26b93538b1f4acddf4bce6961417cbf16/database/products.sql#L1

Reply:
Assuming keeping previous ingredients sounds correct in this context. This will create an ingredient x coffee relationship at a time. You can aggregate ingredients to a coffee whenever the coffee exists. 
I think it makes sense to check for duplicates though and don't let add an ingredient that the coffee already has. 

FIY - my changes to the hashicups-go-client don't let add duplicated ingredients. It will simply update the quantities of the ingredients list based on the original coffee.

I think updating existing coffee_ingredients is out of the context I'm creating this for. I prefer not to do it and leave it for when this becomes necessary. Maybe it never will.   

The idea of the packer plugin is exactly creating new custom coffees out of existing ones, by just updating ingredients quantity, and avoid changing existing coffees from the menu.

Reply:
FYI the method accepts and returns only a single ingredient x coffee relationship. The `model.CoffeeIngredients{}` makes it a bit confusing appearing to be more than one.

Reply:
@sylviamoss  that's a good call out on the naming problem. I agree `product-api-go.data.model.CoffeeIngredients` should be renamed to singular.  Do you plan to handle it in this PR? If not, that's totally cool. We can just create another issue for that.

Reply:
> FIY - my changes to the hashicups-go-client don't let add duplicated ingredients. It will simply add and update the quantities of the ingredients list based on the original coffee. [check it out here](https://github.com/hashicorp-demoapp/hashicups-client-go/pull/1/files#diff-e0034e283fa9acc7419b25ea5cc27986df7e775f56c2cb0b7f369f8e5720b8e6R92)

I'm not sure why that would go in the client. Isn't this domain logic, that we'd want to enforce at the domain layer? I would think the client and the server side handler should be focused on serializing/deserializing across the wire and other net/http concerns, and domain logic would exist in the domain services. 

Regarding this comment - "I think updating existing coffee_ingredients is out of the context I'm creating this for. I prefer not to do it and leave it for when this becomes necessary. Maybe it never will." - I'll reach out and see if we can zoom through this.

Reply:
You know what? I'll move up the custom coffee logic in [hashicups-client](https://github.com/hashicorp-demoapp/hashicups-client-go/pull/1/files#diff-e0034e283fa9acc7419b25ea5cc27986df7e775f56c2cb0b7f369f8e5720b8e6R68) to the packer-plugin-hashicups and leave the client paired with this API. The custom coffee logic is very specific to the packer plugin and I think it's making things more confusing.

Reply:
I made the insert query an upsert one and renamed the struct to CoffeeIngredient

Reply:
Thank you!